IPAdapterPlus.py

```diff
@@ -427,24 +471,37 @@ def apply_ipadapter(self, ipadapter, model, weight, clip_vision=None, image=None
             clip_embed = embeds[0].cpu()
             clip_embed_zeroed = embeds[1].cpu()
         else:
-            if image.shape[1] != image.shape[2]:
-                print("\033[33mINFO: the IPAdapter reference image is not a square, CLIPImageProcessor will resize and crop it at the center. If the main focus of the picture is not in the middle the result might not be what you are expecting.\033[0m")
+            if self.is_face_id:
+                global IPADAPTER_INSIGHT_FACE_APP
+                if IPADAPTER_INSIGHT_FACE_APP is None:
+                    app = FaceAnalysis(name="buffalo_l", providers=['ROCMExecutionProvider', 'CPUExecutionProvider'])
+                    app.prepare(ctx_id=0, det_size=(640, 640))
+                    IPADAPTER_INSIGHT_FACE_APP = app
+
+                faces = IPADAPTER_INSIGHT_FACE_APP.get(tensor2cv(image))
+
+                clip_embed = torch.from_numpy(faces[0].normed_embedding).unsqueeze(0)
```

Comment:
I just use the first face available. Note also that insightface has trouble finding faces sometimes if it can't see the top of the head or more surrounding area. This can sort of be fixed padding the image.